### ISSUE TITLE
DepthwiseConv2D weight gradient violates homogeneity w.r.t upstream gradients

### ISSUE THREAD

**Issue by XiaoHuaOvo-tech** [tensorflow/tensorflow#107706]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TF2.17.0

### Custom code

Yes

### OS platform and distribution

Linux Ubtntu 22.04

### Python version

3.10.19

### CUDA/cuDNN version

CUDA version: 12.3 cuDNN version: 8

### GPU model and memory

Tesla V100

### Current behavior?

Description

We observe a violation of the homogeneity (scaling linearity) property of the weight gradient (gW) in tf.nn.depthwise_conv2d.

Specifically, for a scalar α and upstream gradient G, the following mathematical property is expected to hold for automatic differentiation:

∂W/∂L ​(αG)=α⋅ ∂W/∂L​(G)

However, under TensorFlow’s depthwise_conv2d, the weight gradient does not satisfy this property, even when the upstream gradient is a constant tensor.

This issue is reproducible, systematic, and isolated to the weight gradient path (gW). Input gradients (gX) do not exhibit this behavior.

Expected Behavior

For any scalar α and upstream gradient G:

The gradient w.r.t. convolution weights should be homogeneous:

gW(alpha * G) == alpha * gW(G)

This is a fundamental linearity requirement of reverse-mode automatic differentiation.

Observed Behavior

When scaling the upstream gradient by a scalar α, the computed weight gradient deviates from the expected scaled result beyond reasonable numerical tolerance.

The violation is:

Element-wise (not just norm-based)

Sparse but systematic (affecting a small fraction of elements)

Reproducible across runs

Independent of random upstream gradients (also occurs with constant upstream gradients)

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np

# ------------------------
# vR metric (你的原样代码)
# ------------------------
def compute_vR(A, B, atol=1e-5, rtol=1e-3):
    A_flat = np.asarray(A).flatten()
    B_flat = np.asarray(B).flatten()
    N = len(A_flat)
    diff = np.abs(A_flat - B_flat)
    scale = np.maximum(np.abs(A_flat), np.abs(B_flat))
    tolerance = atol + rtol * scale
    eps = np.finfo(np.float64).tiny
    r_i = diff / (tolerance + eps)
    v_R = np.max(r_i)
    violation = v_R > 1.0
    violation_mask = r_i > 1.0
    rho = np.sum(violation_mask) / N
    E_mean = np.mean(np.maximum(0, r_i - 1.0))
    return v_R, violation, rho, E_mean

# 固定随机性
np.random.seed(0)
tf.random.set_seed(0)

# NHWC
N, H, W, C = 10, 31, 33, 32
kH, kW = 9, 9
depth_multiplier = 1

x = tf.random.normal([N, H, W, C])
kernel = tf.random.normal([kH, kW, C, depth_multiplier])

# ------------------------
# G = 全 1
# ------------------------
with tf.GradientTape() as tape:
    tape.watch(kernel)
    y = tf.nn.depthwise_conv2d(
        x, kernel,
        strides=[1, 1, 1, 1],
        padding="VALID",
        data_format="NHWC"
    )

G = tf.ones_like(y)
gW = tape.gradient(y, kernel, output_gradients=G)

# ------------------------
# G_scaled = alpha * G
# ------------------------
alpha = tf.constant(3.0, dtype=tf.float32)
G_scaled = alpha * G

with tf.GradientTape() as tape2:
    tape2.watch(kernel)
    y2 = tf.nn.depthwise_conv2d(
        x, kernel,
        strides=[1, 1, 1, 1],
        padding="VALID",
        data_format="NHWC"
    )

gW_scaled = tape2.gradient(y2, kernel, output_gradients=G_scaled)

# ------------------------
# vR 判定
# ------------------------
vR, violation, rho, E_mean = compute_vR(
    gW_scaled.numpy(),
    (alpha * gW).numpy()
)

print(f"vR = {vR:.6f}")
print(f"violation = {violation}")
print(f"rho = {rho:.6f}")
print(f"E_mean = {E_mean:.6e}")
```

### Relevant log output

```shell
vR = 3.580573
violation = True
rho = 0.001157
E_mean = 1.954254e-03
```

**Comment by gzmkl** [comment]
There is not problem when alpha <= 1.0. 

However, when alpha > 1.0 (in your script, alpha = 3.0), then your script has a problem: it magnifies the origin gW with a multiple of  alpha.  So in case of alpha > 0, you should compare gW_scaled/alpha and gW.

**Comment by XiaoHuaOvo-tech** [comment]
"Thanks for the comment @gzmkl.
I understand that normalizing by alpha mathematically aligns the scales. However, my concern is about the numerical stability of this specific op in float32.
Magnitude of Error: We are using a  small scalar (alpha=3.0). In float32, such a small scaling factor should not amplify the error to the point where it violates standard tolerances (even when atol=1e-4，rtol=1e-3).
Practical Impact: Since deep learning training relies heavily on float32 (and often accumulates gradients over many steps), if depthwise_conv2d introduces disproportionate noise even with small scale changes, it could lead to potential gradient issues in real-world training.
We believe his indicates that the accumulation logic in depthwise_conv2d might be less robust than other operators in TensorFlow, which is why we flagged it as a precision issue rather than just a test configuration issue.

**Comment by Venkat6871** [comment]
Hi @XiaoHuaOvo-tech ,
Apologies for the delay, and thanks for the detailed report.

I ran your example on the latest TensorFlow nightly ([CPU](https://colab.sandbox.google.com/gist/Venkat6871/0faea5f62ae0b4414cfdd1f41a4e857e/107706_tf_nightly-v-cpu.ipynb) and [GPU](https://colab.sandbox.google.com/gist/Venkat6871/72e5a514d37083bc14610be136142ef4/107706_tf_nightly-v-gpu.ipynb)) and observed the same behavior, with only small numerical differences between devices.
[tf.nn.depthwise_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d) involves large floating-point reductions when computing gradients. Because floating-point addition is not associative (IEEE-754), small differences can appear depending on execution order, especially in reduction-heavy ops. As noted in TensorFlow’s [determinism documentation](https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism), ops may return slightly different results due to changes in the order in which floating-point numbers are added.

Since the results are consistent across CPU and GPU and fall within expected floating-point variance, this appears to be normal numerical behavior rather than a bug in the gradient implementation.
If you are seeing actual training instability compared to a float64 reference, feel free to share more details and we can take a closer look.

Thank you.

**Comment by XiaoHuaOvo-tech** [comment]
Thanks for the reproduction and the explanation about reduction order / IEEE-754 non-associativity — we agree this can introduce numeric variation in reduction-heavy backward kernels.

One clarification: our report comes from a uniform operator-level test harness across many operators/relations/settings. Violations are highly setting-dependent: for a given operator/relation we observe CPU-only, GPU-only, reproducible on both CPU and GPU, or layout-sensitive (NHWC vs NCHW) cases, with severity varying by device/layout.

Our criterion is an element-wise symmetric allclose-style metric:
r_i = |A_i - B_i| / (atol + rtol * max(|A_i|, |B_i|)),
flagging a violation if any r_i > 1. For float32 we use (atol, rtol) = (1e-4, 1e-3), chosen to be relatively permissive and tolerant to benign reduction/rounding noise.Under float32 precision, we consider this a relatively permissive tolerance choice. We do not treat float64 as a strict oracle here, since it still requires dtype-appropriate tolerance calibration.

For `tf.nn.depthwise_conv2d`, your run shows similar behavior on CPU/GPU, which is helpful. Under the same method/tolerances, we also have minimal repro cases (including other operators) where CPU vs GPU differs significantly and/or changing layout yields different violation sets; we can share scripts/logs/env details if useful.

We also observed in a controlled training-level study (Conv2D, TF 2.3.0) that such backward drift can lead to trajectory/behavior differences even when final accuracy changes are small; happy to share details if that context helps.

**Comment by Venkat6871** [comment]
Hi @XiaoHuaOvo-tech  ,
Thank you so much for filing this issue. We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!
